onnxruntime/contrib_ops/cpu/transformers/subgraph_whisper_decoder.cc

```diff
@@ -0,0 +1,274 @@
+  ORT_RETURN_IF(subgraph_outputs[0]->Name() != "logits",
+                "decoder subgraph output 0 shall be named as logits, got: ", subgraph_outputs[0]->Name());
+
+  const ONNX_NAMESPACE::TensorShapeProto* logits_shape = subgraph_outputs[0]->Shape();
+  const ONNX_NAMESPACE::TensorShapeProto* past_shape = subgraph_outputs[first_present_output_index_]->Shape();
+
+  // Save parameters related to the subgraph.
+  ORT_RETURN_IF_ERROR(GetParameters(past_shape, logits_shape, false));
+  num_layers = (static_cast<int>(subgraph_outputs.size()) - first_present_output_index_) / 2;
+
+  // If input_ids's shape is ['batch_size', 1] then use next token as input_ids.
+  // Otherwise in the case of shape ['batch_size', 'sequence'], use sequence as input_ids.
+  const ONNX_NAMESPACE::TensorShapeProto* input_ids_shape = subgraph_inputs[0]->Shape();
+  if (input_ids_shape->dim(1).has_dim_value() && input_ids_shape->dim(1).dim_value() == 1) {
+    use_sequence_as_input_ids_ = false;
+  }
+
+  constexpr auto int32_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_INT32;
+  constexpr auto float32_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT;
+  constexpr auto float16_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT16;
+
+  ORT_RETURN_IF(subgraph_inputs[0]->TypeAsProto()->tensor_type().elem_type() != int32_type,
+                "decoder subgraph input 0 (input_ids) shall have int32 type");
+
+  auto float_type = subgraph_inputs[2]->TypeAsProto()->tensor_type().elem_type();
+  ORT_RETURN_IF(float_type != float32_type && float_type != float16_type,
+                "decoder subgraph input 2 (encoder_hidden_states) shall have float or float16 type");
+
+  for (int i = first_past_input_index_; i < first_past_input_index_ + 4 * num_layers; i++) {
+    ORT_RETURN_IF(subgraph_inputs[i]->TypeAsProto()->tensor_type().elem_type() != float_type,
+                  "decoder subgraph past inputs shall have same data type as that of encoder_hidden_states");
+  }
+
+  for (int i = 0; i < num_subgraph_outputs; i++) {
+    ORT_RETURN_IF(subgraph_outputs[i]->TypeAsProto()->tensor_type().elem_type() != float_type,
+                  "decoder subgraph output shall have same data type as that of encoder_hidden_states");
+  }
+
+  is_output_float16_ = (subgraph_outputs[0]->TypeAsProto()->tensor_type().elem_type() == float16_type);
+
+  return Status::OK();
+}
+
+// Create inputs for decoder from the following data sources:
+// encoder feeds: encoder_input_ids, decoder_input_ids (with start tokens)
+// encoder fetches: logits,
+//                  encoder_hidden_states,
+//                  present_key_self_0, present_value_self_0, ..., present_key_cross_0, present_value_cross_0, ...
+// decoder_feeds: input_ids,
+//                encoder_hidden_states,
+//                present_key_self_0, present_value_self_0, ..., present_key_cross_0, present_value_cross_0, ...
+Status WhisperDecoderSubgraph::CreateInitialFeeds(
```

Comment:
>CreateInitialFeeds [](http://example.com/codeflow?start=31&length=18)

Looks like the difference with T5 decoder subgraph is attention_mask. We may refactor the code later.